qiita_client/plugin.py

```diff
@@ -0,0 +1,330 @@
+        the job identifier, job_parameters is a dictionary with the parameters
+        of the command and output_dir is a string with the output directory
+    required_parameters : dict of {str: (str, list of str)}
+        The required parameters of the command, keyed by parameter name. The
+        values should be a 2-tuple in which the first element is the parameter
+        type, and the second parameter is the list of subtypes (if applicable)
+    optional_parameters : dict of {str: (str, str)}
+        The optional parameters of the command, keyed by parameter name. The
+        values should be a 2-tuple in which the first element is the parameter
+        name, and the second paramater is the default value
+    outputs : dict of {str: str}
+        The description of the outputs that this command generated. The
+        format is: {output_name: artifact_type}
+    default_parameter_sets : dict of {str: dict of {str: str}}
+        The default parameter sets of the command, keyed by parameter set name.
+        The values should be a dictionary in which keys are the parameter names
+        and values are the specific value for each parameter
+
+    Raises
+    ------
+    TypeError
+        If `function` is not callable
+    ValueError
+        If `function` does not accept 4 parameters
+    """
+    def __init__(self, name, description, function, required_parameters,
+                 optional_parameters, outputs, default_parameter_sets=None):
+        self.name = name
+        self.description = description
+
+        # Make sure that `function` is callable
+        if not callable(function):
+            raise TypeError(
+                "Couldn't create command '%s': the provided function is not "
+                "callable (type: %s)" % (name, type(function)))
+
+        # `function` will be called with the following Parameters
+        # qclient, job_id, job_parameters, output_dir
+        # Make sure that `function` can receive 4 parameters
+        if function.__code__.co_argcount != 4:
+            raise ValueError(
+                "Couldn't register command '%s': the provided function does "
+                "not accept 4 parameters (number of parameters: %d)"
+                % (name, function.__code__.co_argcount))
+
+        self.function = function
+        self.required_parameters = required_parameters
+        self.optional_parameters = optional_parameters
+        self.default_parameter_sets = default_parameter_sets
+        self.outputs = outputs
+
+    def __call__(self, qclient, server_url, job_id, output_dir):
+        return self.function(qclient, server_url, job_id, output_dir)
+
+
+class QiitaArtifactType(object):
+    """A Qiita artifact type
+
+        Parameters
+        ----------
+        name : str
+            The artifact type name
+        description : str
+            The artifact type description
+        can_be_submitted_to_ebi : bool
```

Comment:
what's the likelihood that external submission or data exchanges will expand in the future? if it is reasonable, then it may make sense to do a more discoverable structure, something like `external_exchange` which could be `{'ebi': True, 'vamps': False, ...}`. if expansion here is unlikely, then I think the attrs make sense